core/src/test/java/io/questdb/test/griffin/engine/groupby/SampleByTest.java

```diff
@@ -3739,6 +3768,187 @@ public void testSampleByFirstLastWithNonTsOrFilteredSymbolColumn() throws Except
+
+                try (final RecordCursor cursor = factory.getCursor(sqlExecutionContext)) {
+                    assertCursor(expected, cursor, factory.getMetadata(), true);
+                }
+
+                assertFactoryMemoryUsage();
+            }
+        });
+    }
+
+    @Test
+    public void testSampleByFromToFillNull() throws Exception {
+        assertMemoryLeak(() -> {
+            ddl(DDL_FROMTO);
+            drainWalQueue();
+            assertSql(
+                    "ts\tavg\n" +
+                            "2017-12-30T00:00:00.000000Z\t72.5\n" +
+                            "2018-01-04T00:00:00.000000Z\t264.5\n" +
+                            "2018-01-09T00:00:00.000000Z\t432.5\n",
+                    "select ts, avg(x) from fromto\n" +
+                            "sample by 5d"
+            );
+            assertSql(
+                    "ts\tavg\n" +
+                            "2017-12-20T00:00:00.000000Z\tnull\n" +
+                            "2017-12-25T00:00:00.000000Z\tnull\n" +
+                            "2017-12-30T00:00:00.000000Z\t72.5\n" +
+                            "2018-01-04T00:00:00.000000Z\t264.5\n" +
+                            "2018-01-09T00:00:00.000000Z\t432.5\n",
+                    "select ts, avg(x) from fromto\n" +
+                            "sample by 5d from '2017-12-20' fill(null)"
+            );
+            assertSql(
+                    "ts\tavg\n" +
+                            "2017-12-20T00:00:00.000000Z\tnull\n" +
+                            "2017-12-25T00:00:00.000000Z\tnull\n" +
+                            "2017-12-30T00:00:00.000000Z\t72.5\n" +
+                            "2018-01-04T00:00:00.000000Z\t264.5\n" +
+                            "2018-01-09T00:00:00.000000Z\t432.5\n" +
+                            "2018-01-14T00:00:00.000000Z\tnull\n" +
+                            "2018-01-19T00:00:00.000000Z\tnull\n" +
+                            "2018-01-24T00:00:00.000000Z\tnull\n" +
+                            "2018-01-29T00:00:00.000000Z\tnull\n",
+                    "select ts, avg(x) from fromto\n" +
+                            "sample by 5d from '2017-12-20' to '2018-01-31' fill(null)"
+            );
+        });
+    }
+
+    @Test
+    public void testSampleByFromToIsDisallowedForKeyedQueries() throws Exception {
+        assertMemoryLeak(() -> {
+            ddl(DDL_FROMTO);
+            assertException("select ts, avg(x), first(x), last(x), x from fromto\n" +
+                            "where s != '5'\n" +
+                            "sample by 5d from '2017-12-20' to '2018-01-31' fill(42)",
+                    0, "supported");
+            assertException("select ts, avg(x), first(x), last(x), x from fromto\n" +
+                    "where s != '5'\n" +
+                    "sample by 5d from '2017-12-20' to '2018-01-31' fill(42)", 0, "supported");
+        });
+    }
+
+    @Test
+    public void testSampleByFromToNoFill() throws Exception {
```

Comment:
This variation of the test leads to an NPE. Could you check?
```java
    @Test
    public void testSampleByFromToNoFill() throws Exception {
        assertMemoryLeak(() -> {
            ddl(DDL_FROMTO);
            drainWalQueue();
            assertSql(
                    "ts\tavg\n" +
                            "2017-12-30T00:00:00.000000Z\t72.5\n" +
                            "2018-01-04T00:00:00.000000Z\t264.5\n" +
                            "2018-01-09T00:00:00.000000Z\t432.5\n",
                    "select ts1, avg(x) from (select ts as ts1, x from fromto where x > 0)\n" +
                            "sample by 5d from '2017-12-20'"
            );
        });
    }
```

In principle, I'm not sure if we handle nested queries, joins and aliases correctly in `rewriteSampleByFromTo`, so it's worth adding tests with more complex queries + FROM TO clause. WDYT?

Reply:
I'm still getting an NPE when running this test:
```
java.lang.NullPointerException: Cannot read field "token" because "timestamp" is null
	at io.questdb.griffin.SqlOptimiser.rewriteSampleByFromTo(SqlOptimiser.java:4845)
	at io.questdb.griffin.SqlOptimiser.optimise(SqlOptimiser.java:6040)
	at io.questdb.griffin.SqlCompilerImpl.compileExecutionModel0(SqlCompilerImpl.java:1536)
	at io.questdb.griffin.SqlCompilerImpl.compileExecutionModel(SqlCompilerImpl.java:1525)
	at io.questdb.griffin.SqlCompilerImpl.compileUsingModel(SqlCompilerImpl.java:1634)
	at io.questdb.griffin.SqlCompilerImpl.compileInner(SqlCompilerImpl.java:1601)
	at io.questdb.griffin.SqlCompilerImpl.compile(SqlCompilerImpl.java:235)
	at io.questdb.cairo.pool.SqlCompilerPool$C.compile(SqlCompilerPool.java:128)
	at io.questdb.test.tools.TestUtils.printSql(TestUtils.java:1247)
	at io.questdb.test.tools.TestUtils.assertSql(TestUtils.java:605)
	at io.questdb.test.AbstractCairoTest.assertSql(AbstractCairoTest.java:1774)
	at io.questdb.test.griffin.engine.groupby.SampleByTest.lambda$testSampleByFromToNoFill1$28(SampleByTest.java:3903)
```